CNV-21826: Configuring VM for DPDK

sjhala-ccs opened the pull request:
<!--- PR title format: [GH#<gh-issue-id>][BZ#<bz-issue-id>][OCPBUGS#<jira-issue-id>][OSDOCS#<jira-issue-id>]: <short-description-of-the-pr> --->

<!--- If your changes apply to the latest release and/or in-development version of OpenShift, open your PR against the `main` branch.

* For more details about the information requested in this template, see:
  https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/create_or_edit_content.adoc#submit-PR --->

Version(s): 4.13+
<!--- Specify the version or versions of OpenShift your PR applies to. -->

Issue: [CNV-21826](https://issues.redhat.com//browse/CNV-21826)
<!--- Add a link to the Bugzilla, Jira, or GitHub issue, if applicable. --->

Link to docs preview: https://58584--docspreview.netlify.app/openshift-enterprise/latest/virt/virtual_machines/vm_networking/virt-attaching-vm-to-sriov-network.html#virt-configuring-cluster-dpdk_virt-attaching-vm-to-sriov-network
<!--- Add direct link(s) to the exact page(s) with updated content from the preview build. --->

QE review:
- [x] QE has approved this change.
<!--- QE approval is required to merge a PR except for changes that do not impact the meaning of the docs. --->

<!--- Additional information: --->
<!--- Optional: Include additional context or expand the description here.--->

<!--- After you open your PR, ask for review from the OpenShift docs team:
  For community authors: Tag @openshift/team-documentation in a GitHub comment.--->

apinnick commented on modules/virt-configuring-vm-dpdk.adoc:
```diff
@@ -0,0 +1,109 @@
+        - multus:
+            networkName: dpdk-net <7>
+          name: nic-east
+#...
```
Just a small nit, since I will be fixing these in a separate PR. The OCP guidelines say to use `...` instead of `#...` for ellipses in YAML files.

phoracek commented on modules/virt-configuring-vm-dpdk.adoc:
```diff
@@ -0,0 +1,109 @@
+<4> The number of sockets inside the VM. This field must be set to `1` for the CPUs to be scheduled from the same Non-Uniform Memory Access (NUMA) node.
+<5> The number of cores inside the VM. This must be a value greater than or equal to `1`. In this example, the VM is scheduled with 5 hyper-threads or 10 CPUs.
+<6> The size of the huge pages. The possible values for x86-64 architecture are 1Gi and 2Mi. In this example, the request is for 8 huge pages of size 1Gi. 
+<7> The name of the SRI-OV `NetworkAttachmentDefinition` object.
```
```suggestion
<7> The name of the SR-IOV `NetworkAttachmentDefinition` object.
```

phoracek commented on modules/virt-configuring-vm-dpdk.adoc:
```diff
@@ -0,0 +1,109 @@
++
+[source,terminal]
+----
+$ grubby --update-kernel=ALL --args="default_hugepagesz=1GB hugepagesz=1G hugepages=8 isolcpus=2-9"
```
The values use here and in the following commands need to match some of the attributes earlier used in the VM spec, used in other commands of this section (and maybe even in the PerformanceProfile). We should call those out. @orelmisan would you please help defining the mapping?

sjhala-ccs replied:
@orelmisan Do you have any suggested changes for this section?

phoracek replied:
Commented in the thread above. We can close this one.

orelmisan commented on modules/virt-configuring-cluster-dpdk.adoc:
```diff
@@ -0,0 +1,29 @@
+
+.Procedure
+. Create a `SriovNetworkNodePolicy` object with the `spec.deviceType` field set to `vfio-pci`.
+. Create a `RuntimeClass` resource by using the `PerformanceProfile` object of the Node Tuning Operator.
```
@phoracek , I am not aware of a difference between a `PerformanceProfile` for Pods and a `PerformanceProfile` for VMs.

phoracek replied:
Understood, we can close this thread.

apinnick commented on modules/virt-configuring-cluster-dpdk.adoc:
```diff
@@ -0,0 +1,29 @@
+----
+$ oc annotate --overwrite -n openshift-cnv hco kubevirt-hyperconverged \
+    kubevirt.kubevirt.io/jsonpatch='[{"op": "add", "path": "/spec/configuration/defaultRuntimeClass", \
+        "value": <runtimeclass-name>}]'
```
I think line 27 has an extra indent. Shouldn't it align with line 26?

apinnick commented on modules/virt-configuring-cluster-dpdk.adoc:
```diff
@@ -0,0 +1,29 @@
+* The compute nodes have pre-allocated huge pages and Network Interface Cards (NICs) compatible with SR-IOV.
+
+.Procedure
+. Create a `SriovNetworkNodePolicy` object with the `spec.deviceType` field set to `vfio-pci`.
```
Small nit. I think this should be "an", not "a", because AFAIK "SR-IOV" is an initialism (pronounced "ess-ar-eye-oh-vee" etc), not an acronym.

apinnick commented on virt/virtual_machines/vm_networking/virt-attaching-vm-to-sriov-network.adoc:
```diff
@@ -24,7 +24,23 @@ include::modules/nw-sriov-network-attachment.adoc[leveloffset=+1]
+[role="_additional-resources"]
+.Additional resources
+* xref:../../../scalability_and_performance/what-huge-pages-do-and-how-they-are-consumed-by-apps.adoc#configuring-huge-pages_huge-pages[Configuring huge pages]
+* xref:../../../virt/virtual_machines/vm_networking/virt-attaching-vm-to-sriov-network.adoc#nw-sriov-configuring-device_virt-attaching-vm-to-sriov-network[Configuring SR-IOV network devices]
```
I don't think you need this link because the module is included in this assembly.

phoracek commented on modules/virt-configuring-cluster-dpdk.adoc:
```diff
@@ -0,0 +1,32 @@
+* You have installed the Node Tuning Operator.
+* The compute nodes have pre-allocated huge pages and Network Interface Cards (NICs) compatible with SR-IOV.
+
+.Procedure
```
I like that it describes the individual steps clearly. The only issue with this is that it mixes steps that are required for the cluster configuration and those needed for namespace configuration. Could you split the list to make it clear what needs to be done once for the whole cluster and what is required for every new project.

orelmisan replied:
I'll change the place of this step:
> Configure the runtimeClass in the kubevirt CR.

1. Map your compute nodes topology - understand to which NUMA node your SR-IOV NIC is connected to.
2. Label the worker nodes to mark them as DPDK-enabled.
3. Create a MachineConfigPool object to represent the DPDK-enabled worker nodes.
4. Create a PerformanceProfile object with a number of isolated CPUs for your DPDK application from the same NUMA as the NIC (and reserve the rest).
5. Configure the runtimeClass in the kubevirt CR.
6. Create a namespace for your DPDK applications.
7. Create an SriovNetworkNodePolicy object.
8. Create an SriovNetwork object, and configure it to create the NetworkAttachmentDefinition object in the namespace you've just created.
9. Run the VM latency checkup to check that the network is properly configured.
10. Run the DPDK checkup.

phoracek replied:
I did not mean reordering the list, but having two different lists. One would be guide to configure the cluster for cluster admin on day-1, the other would be a guide to configure a namespace for cluster admin (or powerful namespace admin) on day-2

phoracek replied:
We added a suggested changes below. Closing this thread.

phoracek commented on modules/virt-configuring-cluster-dpdk.adoc:
```diff
@@ -0,0 +1,32 @@
+* The compute nodes have pre-allocated huge pages and Network Interface Cards (NICs) compatible with SR-IOV.
+
+.Procedure
+. Create an `SriovNetworkNodePolicy` object with the `spec.deviceType` field set to `vfio-pci`.
+. Create a `RuntimeClass` resource by using the `PerformanceProfile` object of the Node Tuning Operator.
+. Set the container runtime configuration for the `virt-launcher` pod by adding the following annotation to the `HyperConverged` custom resource (CR):
```
```suggestion
. Map your compute nodes topology - understand to which NUMA node your SR-IOV NIC is connected to.
. Label a subset of your worker nodes as DPDK-enabled, e.g. `oc label nodes <node_name> node-role.kubernetes.io/worker-dpdk=""`.
. Create a new `MachineConfigPool` applied to the labeled nodes.
. Create a new `PerformanceProfile` applied to the labeled nodes and the previously configured machine config pool. This profile should select `isolated` CPUs for your DPDK application and `reserved` CPUs for house keeping. Refer to the `PerformanceProfile` documentation for more detail.  
. Create an `SriovNetworkNodePolicy` object with the `spec.deviceType` field set to `vfio-pci`.
. Set the container runtime configuration for the `virt-launcher` pod by adding the following annotation to the `HyperConverged` custom resource (CR):
```

XXX: This may have support implications. We are testing an alternative option. If that option will look ok, we will just need to drop this last step and add a note to the "Create a new `PerformanceProfile`"

phoracek commented on modules/virt-configuring-vm-dpdk.adoc:
```diff
@@ -0,0 +1,109 @@
+// Module included in the following assemblies:
+//
+// * virt/virtual_machines/vm_networking/virt-attaching-vm-to-sriov-network.adoc
+
```
Could we add a new section between cluster and VM configuration?

```suggestion
:_content-type: PROCEDURE
[id="virt-configuring-vm-dpdk_{context}"]
= Configuring a project for DPDK workloads

Prior to running a DPDK-enabled VM, its namespace/project must be configured according.

.Prerequisites
* Your cluster is configured to run DPDK workloads.

.Procedure
. Create a namespace for your DPDK applications.
. Create an `SriovNetwork` object referencing the earlier defined `SriovNetworkNodePolicy`.
. (Optional) Run the VM latency checkup to check that the network is properly configured.
. (Optional) Run the DPDK checkup to verify that the namespace is DPDK-ready.
```

phoracek commented on modules/virt-configuring-vm-dpdk.adoc:
```diff
@@ -0,0 +1,109 @@
+
+.Prerequisites
+* Your cluster is configured to run DPDK workloads.
+* You have created a project on the cluster in which the VM is running.
```
```suggestion
* You have created and configured a project in which the VM will run.
```

phoracek commented on modules/virt-configuring-vm-dpdk.adoc:
```diff
@@ -0,0 +1,109 @@
+          sockets: 1 <4>
+          cores: 5 <5>
+          threads: 2
+          dedicatedCpuPlacement: true
```
There is a BZ we discovered after publishing the upstream docs. We can resolve https://bugzilla.redhat.com/show_bug.cgi?id=2184914 it here (or we can handle it in a follow-up PR).

```suggestion
          dedicatedCpuPlacement: true
          isolateEmulatorThread: true
```

phoracek commented on modules/virt-configuring-vm-dpdk.adoc:
```diff
@@ -0,0 +1,109 @@
+
+. Save and exit the editor.
+. Configure the guest operating system. The following example shows the configuration steps for RHEL8 OS:
+.. Configure isolated CPUs and specify huge pages by using the GRUB bootloader command line interface:
```
```suggestion
.. Using the GRUB bootloader command line interface:
... Configure huge pages. In the example below we use 8 1G huge pages. 
... Select isolated VM's CPUs. In the example below we leave the first two (0 and 1) CPUs for house keeping and isolate the rest for the DPDK application.
```

phoracek commented on modules/virt-configuring-vm-dpdk.adoc:
```diff
@@ -0,0 +1,109 @@
+        cpu-load-balancing.crio.io: disable <1>
+        cpu-quota.crio.io: disable <2>
+        irq-load-balancing.crio.io: disable <3>
+    spec:
```
To make sure we land on the correct node

```suggestion
    spec:
      nodeSelector:
        node-role.kubernetes.io/worker-dpdk: ""
```

sjhala-ccs commented:
@phoracek and @orelmisan I have incorporated your suggestions and added some more details. Please review. Thanks!

Preview build: https://58584--docspreview.netlify.app/openshift-enterprise/latest/virt/virtual_machines/vm_networking/virt-attaching-vm-to-sriov-network.html#virt-configuring-cluster-dpdk_virt-attaching-vm-to-sriov-network

orelmisan commented on modules/virt-configuring-cluster-dpdk.adoc:
```diff
@@ -0,0 +1,117 @@
+  nodeSelector:
+    node-role.kubernetes.io/worker-dpdk: "" 
+----
+
```
After applying these two manifests, there should be a gradual reboot of the relevant nodes.
Please consider adding a note about it.

sjhala-ccs replied:
Does the user initiate the node reboot or does it happen automatically when the MCP and PP manifests are applied?

sjhala-ccs replied:
@orelmisan Does this sound okay: 

> Perform a graceful restart of the compute nodes after you apply the `MachineConfigPool` and `PerformanceProfile`manifests.

I can also link to this procedure if it is relevant: https://docs.openshift.com/container-platform/4.9/nodes/nodes/nodes-nodes-rebooting.html#nodes-nodes-rebooting-gracefully_nodes-nodes-rebooting

orelmisan replied:
> Does the user initiate the node reboot or does it happen automatically when the MCP and PP manifests are applied?

The nodes are automatically rebooted.
I don't know if the user has the ability to manually control this process.

orelmisan commented on modules/virt-configuring-vm-project-dpdk.adoc:
```diff
@@ -0,0 +1,53 @@
+      }],
+      "gateway": "10.56.217.1"
+    }
+  networkNamespace: dpdk-checkup-ns
```
Please consider adding a `<x>` here to explain that this is the namespace you want to deploy the NetworkAttachmentDefinition object to.

orelmisan commented on modules/virt-configuring-vm-project-dpdk.adoc:
```diff
@@ -0,0 +1,53 @@
+* Your cluster is configured to run DPDK workloads.
+
+.Procedure
+. Create a namespace for your DPDK applications:
++
+[source,terminal]
+----
+$ oc create ns dpdk-checkup-ns
+----
```
@phoracek Is there a difference between creating a namespace or creating a project?

phoracek replied:
Frankly, I don't know if OpenShift handles the two differently. I would stick to whatever we use elsewhere in the docs

sjhala-ccs replied:
@orelmisan This is the documentation on projects that I have linked to in the Additional resources section: https://docs.openshift.com/container-platform/4.12/applications/projects/working-with-projects.html

Is it relevant for this procedure?

orelmisan replied:
> @orelmisan This is the documentation on projects that I have linked to in the Additional resources section: https://docs.openshift.com/container-platform/4.12/applications/projects/working-with-projects.html
> 
> Is it relevant for this procedure?

I think so, thanks.

orelmisan reviewed (commented):
Thank you for the changes @sjhala-ccs.
Over all it looks very good.
Please see the inline comments.

phoracek commented on modules/virt-configuring-cluster-dpdk.adoc:
```diff
@@ -0,0 +1,117 @@
+    feature.node.kubernetes.io/network-sriov.capable: "true"
+----
+
+. Set the container runtime configuration for the `virt-launcher` pod by adding the following annotation to the `HyperConverged` custom resource (CR):
```
Sounds good (although I still hope to replace it with the performance profile global)

sjhala-ccs commented on modules/virt-configuring-cluster-dpdk.adoc:
```diff
@@ -0,0 +1,117 @@
++
+[source,terminal]
+----
+$ oc annotate --overwrite -n openshift-cnv hco kubevirt-hyperconverged \
+    kubevirt.kubevirt.io/jsonpatch='[{"op": "add", "path": "/spec/configuration/defaultRuntimeClass", \
+    "value": <runtimeclass-name>}]'
```
I think it renders fine in the preview, the spacing is somehow not showing up correctly in the file view.

Preview:  https://58584--docspreview.netlify.app/openshift-enterprise/latest/virt/virtual_machines/vm_networking/virt-attaching-vm-to-sriov-network.html#virt-configuring-cluster-dpdk_virt-attaching-vm-to-sriov-network

orelmisan replied:
There is a redundant backslash that should be removed.
IIRC it breaks the command.

sjhala-ccs commented on modules/virt-configuring-cluster-dpdk.adoc:
```diff
@@ -0,0 +1,117 @@
+      node-role.kubernetes.io/worker-dpdk: ""
+----
+
+. Generate a performance profile by using the Performance Profile Creator (PPC) command-line tool, which is part of the Node Tuning Operator. The PPC requires the following arguments:
+** `mcp-name`: The machine config pool name corresponding to the target machines. 
+** `reserved-cpu-count`: The number of reserved CPUs for house keeping. 
+** `rt-kernel`: Enable real-time kernel.
```
Okay, but here is the documentation for the PPC in case you are interested: https://docs.openshift.com/container-platform/4.12/scalability_and_performance/cnf-create-performance-profiles.html

orelmisan replied:
Thank you

orelmisan reviewed (approved):
Thank you for the changes @sjhala-ccs

yossisegev commented on modules/virt-configuring-vm-dpdk.adoc:
```diff
@@ -0,0 +1,120 @@
+----
+<1> This annotation specifies that load balancing is disabled for CPUs that are used by the container.
+<2> This annotation specifies that the CPU quota is disabled for CPUs that are used by the container.
+<3> This annotation specifies that Interrupt ReQuest (IRQ) load balancing is disabled for CPUs that are used by the container.
```
"ReQuest" - I understand that it is written this way in order to clarify the IRQ abbreviation, but are you sure it is valid to write it this way (i.e. CamelCase inside a single word)?
If this is indeed a common practice in technical writing, then I apologize in advance for the noise.

sjhala-ccs replied:
Thanks @yossisegev. Our style guidelines do say " When you spell out an abbreviation, do not capitalize letters in the middle of a word to show the origin of an initialism."

I will update the text.

apinnick commented on modules/virt-configuring-vm-dpdk.adoc:
```diff
@@ -0,0 +1,120 @@
+        - multus:
+            networkName: dpdk-net <8>
+          name: nic-east
+...
```
FYI: I updated YAML ellipses in all virt modules, changing `...` to `# ...`.

kelbrown20 commented on modules/virt-configuring-cluster-dpdk.adoc:
```diff
@@ -0,0 +1,123 @@
+[id="virt-configuring-cluster-dpdk_{context}"]
+= Configuring a cluster for DPDK workloads
+
+Use this procedure to configure an {product-title} cluster to run Data Plane Development Kit (DPDK) workloads.
```
```suggestion
You can use the following procedure to configure an {product-title} cluster to run Data Plane Development Kit (DPDK) workloads.
```

kelbrown20 commented on modules/virt-configuring-cluster-dpdk.adoc:
```diff
@@ -0,0 +1,123 @@
+* You have installed the Node Tuning Operator.
+
+.Procedure 
+. Map your compute nodes topology to determine which Non-Uniform Memory Access (NUMA) CPUs will be isolated for DPDK applications and which ones will be reserved for the OS.
```
I know in IBM style we try to avoid the future tense and passive voice as much as we can, which this sentence has a bit of. It still reads fine to me, and I can't think of a suggestion, but I just wanted to make you aware 😄.

sjhala-ccs replied:
Reworded to use present tense

kelbrown20 commented on modules/virt-configuring-cluster-dpdk.adoc:
```diff
@@ -0,0 +1,123 @@
+
+.Procedure 
+. Map your compute nodes topology to determine which Non-Uniform Memory Access (NUMA) CPUs will be isolated for DPDK applications and which ones will be reserved for the OS.
+. Label a subset of the compute nodes with a custom role, for example, `worker-dpdk`:
```
```suggestion
. Label a subset of the compute nodes with a custom role. For example, `worker-dpdk`:
```

kelbrown20 commented on modules/virt-configuring-cluster-dpdk.adoc:
```diff
@@ -0,0 +1,123 @@
+      node-role.kubernetes.io/worker-dpdk: ""
+----
+
+. Create a `PerformanceProfile` manifest that applies to the labeled nodes and the machine config pool that you created in step 2. The performance profile specifies the CPUs that are isolated for DPDK applications and the CPUs that are reserved for house keeping.
```
```suggestion
. Create a `PerformanceProfile` manifest that applies to the labeled nodes and the machine config pool that you generated. The performance profile specifies the CPUs that are isolated for DPDK applications and the CPUs that are reserved for house keeping.
```

kelbrown20 commented on modules/virt-configuring-cluster-dpdk.adoc:
```diff
@@ -0,0 +1,123 @@
++
+[NOTE]
+====
+Adding the annotation to the `HyperConverged` CR changes a global setting that affects all VMs that are created after the annotation is applied.
```
Up to you! 
```suggestion
Adding the annotation to the `HyperConverged` CR changes a global setting that affects all VMs that are created after the annotation applies. 
```

sjhala-ccs replied:
Keeping as is

kelbrown20 commented on modules/virt-configuring-vm-dpdk.adoc:
```diff
@@ -0,0 +1,120 @@
+* You have created and configured the project in which the VM will run.
+
+.Procedure
+. Edit the `VirtualMachine` manifest to include information about the SR-IOV network interface, CPU topology, CRI-O annotations, and huge pages:
```
This is also up to you. I was just thinking when you edit something, it already exists, and just I thought the positive `your` might sounds better.
```suggestion
. Edit your `VirtualMachine` manifest to include information about the SR-IOV network interface, CPU topology, CRI-O annotations, and huge pages:
```

kelbrown20 commented on modules/virt-configuring-vm-dpdk.adoc:
```diff
@@ -0,0 +1,120 @@
+$ oc apply -f <file_name>.yaml
+----
+
+. Configure the guest operating system. The following example shows the configuration steps for RHEL 8 OS:
```
Does this also work for sub versions of RHEL 8? That might be helpful to specify

kelbrown20 commented on modules/virt-configuring-vm-project-dpdk.adoc:
```diff
@@ -0,0 +1,54 @@
+[id="virt-configuring-vm-project-dpdk_{context}"]
+= Configuring a project for DPDK workloads
+
+Configure the project to run DPDK workloads on SR-IOV hardware. 
```
```suggestion
You can configure the project to run DPDK workloads on SR-IOV hardware. 
```

sjhala-ccs commented:
@yossisegev Can you please approve the changes if they look good? I need to get an "LGTM" from you before I can merge the PR. Thank you.

yossisegev commented:
LGTM (CNV network QE)
